### PR TITLE
Prepare for custom channels (lnd v0.18.4-beta)

### PR DESCRIPTION
**Pull request by guggero**
This is the mother of all PRs for the custom channel saga. Merging this PR will be the final step for bringing custom channels to a non-experimental version of litd. There are many dependencies (see list/graph below), so it will remain in draft for a bit.
Most of the code in these PRs have already been reviewed on the `0-19-staging` branch and have just been squashed down into easy-to-review pieces.

The following list is an **ordered** list of dependencies that need to be resolved before this PR can be merged (cc @dstadulis):

 - [x] Review and merge https://github.com/lightninglabs/lndclient/pull/192
 - [x] Review and merge https://github.com/lightningnetwork/lnd/pull/8960
 - [x] Create release branch PR with stable commits (https://github.com/lightningnetwork/lnd/pull/9183)
 - [x] Push tag `lndclient v0.18.4-0` that points to release branch with stable commits
 - [x] Update https://github.com/lightninglabs/pool/pull/488 to point to `lndclient v0.18.4-0` (and with it also bumping the `lnd` version), take out of draft
 - [x] Review and merge https://github.com/lightninglabs/pool/pull/488
 - [x] Update https://github.com/lightninglabs/loop/pull/828 to point to `lndclient v0.18.4-0` (and with it also bumping the `lnd` version), take out of draft
 - [x] Review and merge https://github.com/lightninglabs/loop/pull/828
 - [x] Update https://github.com/lightninglabs/taproot-assets/pull/1130 to point to `lndclient v0.18.4-0` (and with it also bumping the `lnd` version), take out of draft
 - [x] Review and merge https://github.com/lightninglabs/taproot-assets/pull/1130
 - [x] Code, review and merge HTLC sweeping saga:
   - [x] https://github.com/lightningnetwork/lnd/issues/9176
   - [x] https://github.com/lightninglabs/taproot-assets/pull/1154
   - [x] https://github.com/lightningnetwork/lnd/pull/9194
 - [x] Wait for `lnd v0.18.4-beta` to be tagged (at least an RC tag) https://github.com/lightningnetwork/lnd/issues/9173
 - [x] https://github.com/lightninglabs/lndclient/issues/194 point to the tagged version of `lnd`
 - [x] Push tag `lndclient v0.18.4-6`
 - [x] ~Create and merge Pool PR to reference `lndclient v0.18.4-6`~ (not required, `litd` compiles without)
 - [x] ~Create and merge Loop PR to reference `lndclient v0.18.4-6`~ (not required, `litd` compiles without)
 - [x] Create and merge Taproot Assets PR to reference `lndclient v0.18.4-1`
 - [x] Update this PR to point to all the above projects and their merged commits (`loop`, `pool`, `tapd`)

NOTE: This PR includes all recent changes to `litd`'s `0-19-staging` branch, including the new liquidity tests from https://github.com/lightninglabs/lightning-terminal/pull/842. There haven't been any other in-flight PRs to the staging branch when writing this.

### CODE REVIEW

**Comment by guggero** [comment]
Similar to https://github.com/lightninglabs/loop/pull/828, I bumped to `lndclient v0.18.4-0` which references https://github.com/lightningnetwork/lnd/pull/9183 (commit https://github.com/lightningnetwork/lnd/pull/9183/commits/ca3bde901eb86638256a72fd17b75ce5951d119f) which should remain stable, so we can start the main review work here.

We will want to wait with merging this until the full dependency chain above is resolved. But once `lnd v0.18.4-beta` is tagged, that should only entail bumping the versions in `go.mod`.

**Comment by guggero** [comment]
Need to add https://github.com/lightninglabs/lightning-terminal/pull/861.

**Comment by guggero** [comment]
@jamaljsr I tagged you for review on this with the intention of asking you to run a full manual test run of this version within Polar (all the tests you've run previously).
Do you think you'll have time for that this or next week? Thank you!

**Comment by jamaljsr** [comment]
@guggero Thanks for the clarification. I wasn't sure how I should test this or if it was ready. I will try it out in Polar by early next week at the latest, hopefully sooner. Excited to see this functionality getting mainlined 🙌

File: config.go
Comment (ellemouton) [inline config.go:195]: nice!

File: subservers/taproot-assets.go
Comment (ellemouton) [inline subservers/taproot-assets.go:109]: 👍  the big switch 💡

File: subservers/faraday.go
Comment (ellemouton) [inline subservers/faraday.go:130]: should it then instead return an `fn.Option`?

File: subservers/manager.go
Comment (ellemouton) [inline subservers/manager.go:87]: adding [this PR](https://github.com/lightninglabs/lightning-terminal/pull/877) which changes the storage of subservers to be a map indexed by name. Then we dont need to iterate here.

File: config.go
Comment (ellemouton) [inline config.go:630]: if tapd is disabled, maybe we should not force this? 

(note to self that maybe we can eventually hide this type of thing behind the `SubServer` interface so that we can be sure that any config tweaks only happen if a sub-server is active. But yeah - i'll take care of that refactor)

File: terminal.go
Comment (ellemouton) [inline terminal.go:514]: yeah another point for hiding this behind the `SubServer` interface (but again - i'll do that)

File: cmd/litcli/main.go
Comment (ellemouton) [inline cmd/litcli/main.go:302]: naming nit: this is a `litConn` no?

File: cmd/litcli/main.go
Comment (ellemouton) [inline cmd/litcli/main.go:88]: are these used?

File: cmd/litcli/main.go
Comment (ellemouton) [inline cmd/litcli/main.go:351]: i wonder if we can't rather expand on the new BakeSuperMacaroon functionality?

Comment (ellemouton) [reply cmd/litcli/main.go:351]: we can just edit the proto message such that it takes an optional sub-server name

File: .github/workflows/main.yml
Comment (ellemouton) [inline .github/workflows/main.yml:247]: thank youuuuuuuu

**Review by ellemouton — COMMENTED** [review]
Looking good!

File: subservers/manager.go
Comment (guggero) [reply subservers/manager.go:87]: Okay, will rebase this PR once the other one is in and then update this.

File: config.go
Comment (guggero) [reply config.go:630]: Hmm, at this point in parsing the config we haven't really done any validation for the taproot asset mode yet, so not sure it's wise to access that here?
Also, this isn't really super consequential, it just allows an RPC user to send p2p error messages to the peers. So no big deal if this is enabled IMO.

File: cmd/litcli/main.go
Comment (guggero) [reply cmd/litcli/main.go:88]: Yes, indirectly. Added a comment above.

File: cmd/litcli/main.go
Comment (guggero) [reply cmd/litcli/main.go:351]: Ah, cool, baking a super macaroon works well here. Sure, we could narrow the permissions later, but for now this is great.

**Comment by guggero** [comment]
This PR now contains all the recently merged PRs of the `0-19-staging` branch up to commit 2194f1e8.

File: subservers/manager.go
Comment (guggero) [reply subservers/manager.go:99]: Updated.

**Review by jamaljsr — APPROVED** [review]
tACK LGTM 🔥 🚀

I tested the following scenarios in Polar using this network structure with all `litd` nodes except for `erin` which is Core Lightning v24.08.1.

```
alice --[asset]-> bob --[sats]-> carol --[asset]-> dave
                                   |
                                 [sats]
                                   |
                                   v
                                  erin (core-ln)
```

- fund asset channel using full asset balance after mint
- fund asset channel using partial asset balance after mint
- fund asset channel using full asset balance after mint + onchain send
- fund asset channel using partial asset balance after mint + onchain send
- fund asset channel using full asset balance from onchain receive
- fund asset channel using partial asset balance from onchain receive
- fund asset channel with `push_sat=5000`
- create asset invoice on dave, pay with asset on alice (multi-hop)
- create asset invoice on bob, pay with asset on alice (direct peer)
- create asset invoice on alice, pay with asset on bob (direct peer, opposite direction)
- create asset invoice on alice, pay with asset on dave (multi-hop, opposite direction)
- create asset invoice on dave, pay with sats on bob (sats -> asset)
- create sats invoice on bob, pay with asset on dave (asset -> sats)
- create asset invoice on alice, pay with sats on erin (cln sats -> asset)
- create sats invoice on erin, pay with asset on alice (asset -> cln sats)
- create sats invoice on bob, pay with sats on alice (direct peer, sats -> sats)
- create sats invoice on dave, pay with sats on alice (multi-hop, sats -> sats)
- send asset keysend payment from alice to bob
- send asset keysend payment from alice to carol

Note: I was using commit [35e6aea](https://github.com/lightninglabs/lightning-terminal/commit/35e6aea69520a497a3896276286009cf4f0a2bd6). I see there have been more pushes since I started testing. Let me know if you'd like another round of testing using the final commits before merging.

**Comment by guggero** [comment]
Thank you very much for testing, @jamaljsr!!! :pray: 

Yes, I did add some more commits. Most of them are fixes for the CI and Go module bumps.
But because the final merged `tapd` PR (https://github.com/lightninglabs/taproot-assets/pull/1130) was on top of the RFQ changes, some rounding and precision related calculations now work slightly differently (e.g. with _better_ precision and _fewer_ rounding losses).

So I don't think a full test suite run is necessary. But perhaps some of the multi-hop scenarios with bigger asset values would be beneficial. Thank you!

This PR now contains all changes of the staging branch, most notably the following PRs (and everything before those):
 - https://github.com/lightninglabs/lightning-terminal/pull/871
 - https://github.com/lightninglabs/lightning-terminal/pull/865
 - https://github.com/lightninglabs/lightning-terminal/pull/861
 - https://github.com/lightninglabs/lightning-terminal/pull/842
 - https://github.com/lightninglabs/lightning-terminal/pull/838

**Comment by jamaljsr** [comment]
> So I don't think a full test suite run is necessary. But perhaps some of the multi-hop scenarios with bigger asset values would be beneficial. Thank you!

I've been testing large asset payments and came across something that I'm not sure is expected behavior, so I figured I flag it just in case. 

I am using the same network structure I mentioned in https://github.com/lightninglabs/lightning-terminal/pull/848#pullrequestreview-2385209814 except now the assets channels have 10M PUSD and the sats channel has 10BTC. When I make a 5M PUSD multi-hop invoice payment from `alice` to `dave`, the resulting channel asset balances between `alice` and `bob` has 4,999,990 local and 5,000,010 remote. Whereas, the channel asset balances between `carol` and `dave` are 5,000,000 on both local & remote as expected.

Is there a reason why the asset balances are slightly off in the first channel? Are fees paid for using the asset instead of sats and factored into the exchange rate?

BTW, I'm using `--taproot-assets.experimental.rfq.mockoracleassetsperbtc=1000000` so there is a 100 sats to 1 PUSD exchange rate.

Here are the logs from all 4 nodes: [polar_large_pmt_logs.zip](https://github.com/user-attachments/files/17508449/polar_large_pmt_logs.zip). 

I also took a screen capture of the behavior in Polar, if it's helpful.

https://github.com/user-attachments/assets/434d5884-2b8f-47b5-b6ab-686ecb9282ac

**Comment by guggero** [comment]
Thank you for flagging, @jamaljsr.
This is indeed correct. The routing fees are paid with assets. By default, the fees are 1ppm (and 1 sat base fee) for each hop, so things add up. 1ppm of 5 million units is 5 units, with 2 hops that's 10 units. The base fee is truncated because of the inaccuracy of the asset unit (1 unit represents multiple satoshi instead of the other way around).

Coincidentally I just finished a PR today that does the conversion rate and balance accounting down to the single satoshi in https://github.com/lightninglabs/lightning-terminal/pull/883. So I'm quite confident everything is correct now and some of the previous inaccuracies of the RFQ system have been resolved.

**Comment by guggero** [comment]
Of course the routing fees are actually paid in sats to the two routing nodes, calculated from the satoshi amount of the forwarded payment. But _expressed_ in asset units, those fees add up to 10 units for the specific payment.

**Comment by jamaljsr** [comment]
Thanks so much for the explanation. Now it totally makes sense. 

I just have one follow-up question, for my own information, in the cases when sending smaller asset payments. If I only send 500 PUSD, how are the routing fees distributed? I assume the equivalent sats value is too small to be represented as an asset unit. Are the fees then added to the sats amount of the HTLC?

**Comment by guggero** [comment]
There is a certain amount of "underpayment" allowed when the conversion from asset to sats is off due to the precision. Currently 1 asset unit per HTLC (and up to 16 asset units per invoice, due to the MPP default max shard amount of 16) is allowed. That precision loss is basically paid by the edge node.
To avoid those precision losses to be significant amounts, the asset precision (decimal display) should be chosen correctly (a value of 6 for example, see the RFQ doc in the `tapd` repo).

**Comment by ViktorTigerstrom** [comment]
See:
https://github.com/lightninglabs/lightning-terminal/issues/907

**Comment by jamaljsr** [comment]
I tried to build the latest commit ([dff9b28](https://github.com/lightninglabs/lightning-terminal/pull/848/commits/dff9b28ba742f123e5ac08571fed2d328c2c37e7)) today and ran into this error
```
$ make install
go: downloading github.com/lightninglabs/aperture v0.3.2-beta.0.20241015115230-d59b5514c19a
/Users/jamal/gocode/pkg/mod/github.com/lightninglabs/loop@v0.28.8-beta.0.20241022072406-1e8ae31ddc27/client.go:15:2: github.com/lightninglabs/aperture@v0.3.2-beta.0.20241015115230-d59b5514c19a: invalid version: unknown revision d59b5514c19a
```

**Comment by guggero** [comment]
Thanks for the report. Looks like a temporary commit of aperture is no longer reachable but was still cached for me.
Fixed it by pushing an updated loop version. Hope it works for you as well now.

**Comment by jamaljsr** [comment]
Thanks for the quick turnaround. Unfortunately I'm still getting a similar build error with the updated loop version when using commit [27316c8](https://github.com/lightninglabs/lightning-terminal/pull/848/commits/27316c8281be5bb1036d1cf60254ff27378cbefc)

```
go install -v -trimpath -tags="litd autopilotrpc signrpc walletrpc chainrpc invoicesrpc watchtowerrpc neutrinorpc peersrpc" -ldflags " -X github.com/lightningnetwork/lnd/build.Commit=lightning-terminal-v0.14.0-alpha.rc1-1-g27316c8281be5bb1036d1cf60254ff27378cbefc -X github.com/lightningnetwork/lnd/build.CommitHash=27316c8281be5bb1036d1cf60254ff27378cbefc -X github.com/lightningnetwork/lnd/build.GoVersion= -X github.com/lightningnetwork/lnd/build.RawTags=litd,autopilotrpc,signrpc,walletrpc,chainrpc,invoicesrpc,watchtowerrpc,neutrinorpc,peersrpc -X github.com/lightninglabs/lightning-terminal.appFilesPrefix= -X github.com/lightninglabs/lightning-terminal.Commit=v0.14.0-alpha.rc1-1-g27316c8281be5bb1036d1cf60254ff27378cbefc -X github.com/lightninglabs/loop.Commit=v0.28.9-beta.0.20241121150838-d198bb8230a8 -X github.com/lightninglabs/pool.Commit=v0.6.5-beta.0.20241015105339-044cb451b5df -X github.com/lightninglabs/taproot-assets.Commit=v0.5.0-rc1" github.com/lightninglabs/lightning-terminal/cmd/litd
go: downloading github.com/lightninglabs/aperture v0.3.2-beta.0.20241015115230-d59b5514c19a
/root/go/pkg/mod/github.com/lightninglabs/loop@v0.28.9-beta.0.20241121150838-d198bb8230a8/client.go:15:2: github.com/lightninglabs/aperture@v0.3.2-beta.0.20241015115230-d59b5514c19a: invalid version: unknown revision d59b5514c19a
Makefile:145: recipe for target 'go-install' failed
make: *** [go-install] Error 1
```

**Comment by guggero** [comment]
Okay, I see what's going on. The wrong version comes from `taproot-assets`. Pushed a temporary replace, will create a permanent fix in `tapd`.

**Comment by jamaljsr** [comment]
Yup, that resolved the issue. Thank you for the fix 🙏

**Review by ellemouton — COMMENTED** [review]
LGTM from a Lit standpoint! I do think we should get a tap-asset engineer to ACK the itests and commands specifically though 🙏 

Holding back on the green check until LND 18.4 is out of RC phase

File: subservers/taproot-assets.go
Comment (guggero) [reply subservers/taproot-assets.go:262]: It can _run_ in remote mode, yes. But then the taproot asset channel functionality won't be enabled (for that both `lnd` and `tapd` must be running in integrated mode).

File: subservers/taproot-assets.go
Comment (ViktorTigerstrom) [inline subservers/taproot-assets.go:110]: Should we mention this requirement in the `README`, maybe under the compatibility section?

File: cmd/litcli/ln.go
Comment (ViktorTigerstrom) [inline cmd/litcli/ln.go:253]: Could maybe be worth `nil` checking the result of `resp.GetPaymentResult()` before returning here?

File: cmd/litcli/ln.go
Comment (ViktorTigerstrom) [inline cmd/litcli/ln.go:292]: nit:
```suggestion
		return fmt.Errorf("unable to make rpc conn: %w", err)
```

File: itest/litd_node.go
Comment (ViktorTigerstrom) [inline itest/litd_node.go:659]: nit:
Line length exceeds 80 chars.

File: itest/litd_accounts_test.go
Comment (ViktorTigerstrom) [inline itest/litd_accounts_test.go:468]: nit:
Line length exceeds 80 chars.

File: itest/litd_custom_channels_test.go
Comment (ViktorTigerstrom) [inline itest/litd_custom_channels_test.go:1]: nit:
Line length exceeds 80 characters in multiple occurrences in this file.

File: README.md
Comment (ViktorTigerstrom) [inline README.md:78]: Just leaving a note that this commit is not the last commit in the PR. I understand that this is intended to be fixed later though :)

**Review by ViktorTigerstrom — COMMENTED** [review]
LGTM from a LiT standpoint from me as well :rocket:! Will do final testing and hit approve once `lnd 0.18.4` is released!

File: go.mod
Comment (guggero) [reply go.mod:21]: Hmm, actually, there seems to be a completely new `loopd` version out anyway, so I can bump both.

**Pull request merged**
